community/planning/publisher_for_scabbard_v0.7.md

```diff
@@ -0,0 +1,484 @@
+                .map_err(|_| InternalError)?
+            {
+                // Did not expect any results
+                Ok(Some(_)) => Err(InternalError{..}),
+                Ok(None) => Ok(()),
+                Err(_) => Err(InternalError{..}),
+            }
+        } else {
+            // already called finish or cancel
+            Err(InternalError{..})
+        }
+    }
+}
+
+impl<B: Batch, C: PublisherContext, R: PublishedResult> Drop for PublishHandle<B, C, R> {
+    fn drop(&mut self) {
+        if let Some(sender) = self.sender.take() {
+            match sender.send(PublishMessage::Dropped) {
+                Ok(_) => (),
+                Err(_) => {
+                    error!("Unable to shutdown Publisher thread")
+                }
+            }}
+    }
+}
+
+```
+
+### Example
+An example of using the above API is as follows:
+
+```rust
+use one_batch::{
+        BatchContext, BatchIter, OneBatch, OneBatchVerifierFactory, PublishBatchResult,
+        PublishBatchResultCreatorFactory,
+};
+
+let result_creator_factory = Box::new(PublishBatchResultCreatorFactory::new());
+let batch_verifier_factory = Box::new(OneBatchVerifierFactory::new());
+let mut publisher_starter: PublishFactory<OneBatch, BatchContext, PublishBatchResult> =
+        PublishFactory::new(result_creator_factory, batch_verifier_factory);
+
+let pending_batches = Box::new(BatchIter::new());
+
+let context = BatchContext::new(
+    "test_circuit".to_string(),
+    "test_service".to_string(),
+    "abcd".to_string(),
+);
+
+println!("Starting Publish");
+let publisher_finisher = publisher_starter.start(context, pending_batches)?;
+
+let result = publisher_finisher.finish()?;
+println!("Results {:?}", result);
+```
+
+## Drawbacks
+The current publishing thread contains a very tight loop. One solution would be
+to move the checking for a pending batch into a separate thread and send it to
+the main thread over a sender. This requires adding a new `PublishMessage`
+variant for `PublishMessage::NewBatch`. This would also require that the
+`PendingBatches` implementation would block if no batches are currently
+available, only returning `None` if no more batches will be provided. It is
+important to note that this requires yet another thread for each active service
+which may not be desirable.
+
+Another option would be to add a timeout to the `PendingBatches.next()`
+function that would cause the pending batch to block for some specified
+duration. If a new batch is available it would be returned right away, if not,
+after the duration None would be returned. At this point the `PublishMessage`
+receiver would be checked to see if any finish/cancel messages have been
+received. This would cause some amount of delay between when the
+`PublishMessage` will be received and handled. On the plus side this would not
+require another thread.
+
+## Rationale and alternatives
+
+Another option investigated was to define the Publisher as a set of structs that
```

Comment:
I think "structs" was meant to be "traits" here

Reply:
Fixed